app/src/main/java/it/niedermann/owncloud/notes/model/SingleNoteWidgetFactory.java

```diff
@@ -48,57 +54,56 @@ public int getViewTypeCount() {
 
     @Override
     public void onDataSetChanged() {
+        SharedPreferences sharedprefs = PreferenceManager.getDefaultSharedPreferences(mContext);
+        long noteID = sharedprefs.getLong(SingleNoteWidget.WIDGET_KEY + mAppWidgetId, -1);
 
+        if (noteID >= 0) {
+            if (db.getNote(noteID) == null) {
+                Log.e(TAG, "Error: note not found");
+                note = null;
+            } else {
+                note = db.getNote(noteID);
```

Comment:
If the note exists (which is the default scenario) `db.getNote(noteID)` is executed twice. Since database queries are expensive, this should be optimized.